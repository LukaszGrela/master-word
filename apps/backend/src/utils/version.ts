import fs from 'fs/promises';

export const getVersion = async () => {
  const filePath = './package.json';
  if (process.env.NODE_ENV === 'development') {
    return 'Local dev';
  }
  try {
    const data = await fs.readFile(filePath);
    const packagejson = JSON.parse(data.toString());

    return packagejson.version;
  } catch (error) {
    console.log(error);
    return 'version not found';
  }
};
