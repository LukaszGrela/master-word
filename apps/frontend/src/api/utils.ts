import { TPartialGameState } from './types';

export const createGameState = (
  attempts: number,
  wordLength: number
): TPartialGameState => {
  const word = Array.from(Array(wordLength)).map(() => '');
  return Array.from(Array(attempts)).map(() => ({
    word: word.concat(),
  }));
};
