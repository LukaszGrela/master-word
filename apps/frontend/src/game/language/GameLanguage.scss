@import '../../styles/variables';

.game-language {
  button.translation-btn {
    display: flex;
    gap: 0.25rem;
    color: inherit;

    @media (prefers-color-scheme: dark) {
      background-color: $language-btn-background-dark;
      border-color: $game-language-btn-border-dark;
    }
    &:hover {
      border-color: $language-btn-border-hover;
      @media (prefers-color-scheme: dark) {
        border-color: $game-language-btn-border-hover-dark;
        background-color: $game-language-btn-background-hover-dark;
      }
    }
    &.selected {
      @media (prefers-color-scheme: dark) {
        background-color: $game-language-btn-background-dark;
        border-color: $game-language-btn-selected-border-dark;
      }
    }
  }
}
