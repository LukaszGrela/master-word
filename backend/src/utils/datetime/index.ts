export const getStartOfDay = (day = new Date()): Date => {
  return new Date(day.getFullYear(), day.getMonth(), day.getDate(), 0, 0, 0, 0);
};
