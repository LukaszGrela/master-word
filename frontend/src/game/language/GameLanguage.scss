.game-language {
  .translation-btn {
    display: flex;
    gap: 0.25rem;

    color: inherit;
  }
}
