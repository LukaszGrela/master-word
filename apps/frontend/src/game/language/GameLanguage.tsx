import { FC, useMemo, useState } from 'react';
import { EStorageKeys, AppStorage } from '@repo/utils';
import { getFlag } from '@repo/shared-ui';
import LanguageSelector from './LanguageSelector';
import { useLanguage } from '../../i18n';
import { useConfig } from '../../config';
import { ILanguageListItem } from './types';

import './GameLanguage.scss';

const GameLanguage: FC = () => {
  const { config } = useConfig();
  const storage = AppStorage.getInstance();
  const { getUIText: t } = useLanguage();

  const [wordLanguage, setWordLanguage] = useState<string | undefined>(
    storage.getItem(EStorageKeys.GAME_LANGUAGE) ||
      storage.getItem(EStorageKeys.UI_LANGUAGE) ||
      'pl',
  );

  const handleWordLanguageChange = (language: string): void => {
    AppStorage.getInstance().setItem(EStorageKeys.GAME_LANGUAGE, language);
    setWordLanguage(language);
  };

  const languages = useMemo(() => {
    return config.enabledLanguages.map(
      (language) =>
        ({
          value: language,
          label: t(`word-language-selector-btn-label-${language}`),
          title: t(`result-language-selector-btn-title-${language}`),
          icon: getFlag(language),
        }) as ILanguageListItem,
    );
  }, [config.enabledLanguages, t]);

  return (
    <div className="game-language">
      {languages.length > 0 && (
        <>
          <p className="language-info">{t('result-language-info')}</p>
          <LanguageSelector
            list={languages}
            language={wordLanguage}
            onClick={handleWordLanguageChange}
          />
        </>
      )}
    </div>
  );
};

export default GameLanguage;
