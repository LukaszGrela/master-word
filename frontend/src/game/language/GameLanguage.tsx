import { FC, useState } from 'react';
import LanguageSelector from './LanguageSelector';
import { useLanguage } from '../../i18n';
import { TSupportedLanguages } from '../../api';
import { EStorageKeys } from '../../utils/localStorage/enums';
import { AppStorage } from '../../utils/localStorage';

import './GameLanguage.scss';

const GameLanguage: FC = () => {
  const storage = AppStorage.getInstance();
  const { getUIText: t } = useLanguage();

  const [wordLanguage, setWordLanguage] = useState<
    TSupportedLanguages | undefined
  >(
    (storage.getItem(EStorageKeys.GAME_LANGUAGE) ||
      storage.getItem(EStorageKeys.UI_LANGUAGE) ||
      'pl') as TSupportedLanguages
  );

  const handleWordLanguageChange = (language: TSupportedLanguages): void => {
    AppStorage.getInstance().setItem(EStorageKeys.GAME_LANGUAGE, language);
    setWordLanguage(language);
  };

  return (
    <div className='game-language'>
      <p className='language-info'>{t('result-language-info')}</p>
      <LanguageSelector
        language={wordLanguage}
        onClick={handleWordLanguageChange}
        translationOverride={{
          screenReaderInfo: t('result-language-selector-sr'),
          buttonTitles: {
            en: t('result-language-selector-btn-title-en'),
            pl: t('result-language-selector-btn-title-pl'),
          },
          buttonLabels: {
            en: t('word-language-selector-btn-label-en'),
            pl: t('word-language-selector-btn-label-pl'),
          },
        }}
      />
    </div>
  );
};

export default GameLanguage;
