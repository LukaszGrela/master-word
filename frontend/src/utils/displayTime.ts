import { toHMS } from './datetime/toHMS';

export const displayTime = (ms: number): string => {
  const hms = toHMS(ms);

  if (!hms.h && !hms.m) {
    return `${hms.s.toFixed(3)}s`;
  }

  if (!hms.h && hms.m !== undefined) {
    return `${hms.m}m ${hms.s}s`;
  }

  if (hms.h !== undefined && hms.m !== undefined) {
    return `${hms.h}h ${hms.m}m ${hms.s}s`;
  }

  return '';
};
