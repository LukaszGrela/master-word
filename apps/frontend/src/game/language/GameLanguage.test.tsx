import { Mock, afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, waitFor } from '@testing-library/react';
import {
  contextRenderer,
  defaultLanguageContext,
} from '../../i18n/LanguageContext/__tests__/contextRenderer';
import {
  defaultConfigContext,
  MockContext as MockConfigContext,
} from '../../config/__tests__/contextRenderer';
import { AppStorage, IStorage } from '@repo/utils';
import GameLanguage from './GameLanguage';

vi.mock('@repo/utils', async (importOriginal) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
  const original = await importOriginal<any>();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const AppStorageMock = vi.fn<any, IStorage>();
  const proto = AppStorageMock.prototype as IStorage;

  proto.getItem = vi.fn();
  proto.setItem = vi.fn();
  proto.removeItem = vi.fn();
  proto.has = vi.fn();
  proto.clear = vi.fn();
  proto.clearStorage = vi.fn();

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const instance = new AppStorageMock();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
  (AppStorageMock as any).getInstance = vi.fn().mockReturnValue(instance);

  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    ...original,

    AppStorage: AppStorageMock,
  };
});

describe('game', () => {
  describe('language', () => {
    describe('GameLanguage', () => {
      let storage: IStorage = AppStorage.getInstance();
      beforeEach(() => {
        storage = AppStorage.getInstance();
      });
      afterEach(() => {
        vi.clearAllMocks();
      });
      it('Renders', () => {
        const { container } = contextRenderer(
          <MockConfigContext>
            <GameLanguage />
          </MockConfigContext>,
        );

        expect(container).toMatchSnapshot();
      });
      it('Renders with context', () => {
        const { container } = contextRenderer(
          <MockConfigContext>
            <GameLanguage />
          </MockConfigContext>,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation((translationId: string) => {
                switch (translationId) {
                  case 'translation-button-pl':
                    return 'BTN PL';
                  case 'translation-button-en':
                    return 'BTN EN';
                  case 'translation-info-sr':
                    return 'screen reader info';
                  default:
                    return translationId;
                }
              }),
            },
            renderOptions: {},
          },
        );

        expect(container).toMatchSnapshot();
      });
      it('Renders with context - game language en', async () => {
        (
          storage.getItem as Mock<[key: string], string | null>
        ).mockImplementation((key) => {
          if (key === 'word-language') {
            return 'en';
          }
          return null;
        });
        const mockSetItem = (
          storage.setItem as Mock<[key: string, value: string], void>
        ).mockImplementation(() => {});

        const { container, getByTitle } = contextRenderer(
          <MockConfigContext
            context={{
              ...defaultConfigContext,
              config: {
                ...defaultConfigContext.config,
                enabledLanguages: ['pl', 'en', 'fr'],
              },
            }}
          >
            <GameLanguage />
          </MockConfigContext>,
          {
            languageContext: {
              ...defaultLanguageContext,
              getUIText: vi.fn().mockImplementation((translationId: string) => {
                switch (translationId) {
                  case 'translation-button-pl':
                    return 'BTN PL';
                  case 'translation-button-en':
                    return 'BTN EN';
                  case 'translation-info-sr':
                    return 'screen reader info';
                  case 'result-language-selector-btn-title-pl':
                    return 'GAME LANG PL';
                  case 'result-language-selector-btn-title-en':
                    return 'GAME LANG EN';
                  case 'result-language-selector-btn-title-fr':
                    return 'GAME LANG FR';
                  default:
                    return translationId;
                }
              }),
            },
            renderOptions: {},
          },
        );

        expect(container).toMatchSnapshot();

        const option1 = await waitFor(() => getByTitle('GAME LANG EN'));
        fireEvent.click(option1);

        // it is already selected, should not call callback
        expect(mockSetItem).not.toBeCalled();

        const option2 = await waitFor(() => getByTitle('GAME LANG FR'));

        fireEvent.click(option2);

        await waitFor(() => {
          expect(mockSetItem).toBeCalledWith('word-language', 'fr');
        });
      });
    });
  });
});
