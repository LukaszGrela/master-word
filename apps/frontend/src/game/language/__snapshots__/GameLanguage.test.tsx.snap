// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > language > GameLanguage > Renders 1`] = `
<div>
  <div
    class="game-language"
  >
    <p
      class="language-info"
    />
    <div
      class="language-selector translation"
    >
      <span
        class="hidden sr"
      />
      <button
        class="translation-btn selected"
      >
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/pl.png"
            srcset="https://flagcdn.com/w40/pl.png 2x"
            width="20"
          />
        </span>
      </button>
      <button
        class="translation-btn"
      >
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/us.png"
            srcset="https://flagcdn.com/w40/us.png 2x"
            width="20"
          />
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`game > language > GameLanguage > Renders with context - game language en 1`] = `
<div>
  <div
    class="game-language"
  >
    <p
      class="language-info"
    >
      result-language-info
    </p>
    <div
      class="language-selector translation"
    >
      <span
        class="hidden sr"
      >
        screen reader info
      </span>
      <button
        class="translation-btn"
        title="GAME LANG PL"
      >
        <span
          class="label"
        >
          word-language-selector-btn-label-pl
        </span>
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/pl.png"
            srcset="https://flagcdn.com/w40/pl.png 2x"
            width="20"
          />
        </span>
      </button>
      <button
        class="translation-btn selected"
        title="GAME LANG EN"
      >
        <span
          class="label"
        >
          word-language-selector-btn-label-en
        </span>
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/us.png"
            srcset="https://flagcdn.com/w40/us.png 2x"
            width="20"
          />
        </span>
      </button>
      <button
        class="translation-btn"
        title="GAME LANG FR"
      >
        <span
          class="label"
        >
          word-language-selector-btn-label-fr
        </span>
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/fr.png"
            srcset="https://flagcdn.com/w40/fr.png 2x"
            width="20"
          />
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`game > language > GameLanguage > Renders with context 1`] = `
<div>
  <div
    class="game-language"
  >
    <p
      class="language-info"
    >
      result-language-info
    </p>
    <div
      class="language-selector translation"
    >
      <span
        class="hidden sr"
      >
        screen reader info
      </span>
      <button
        class="translation-btn selected"
        title="result-language-selector-btn-title-pl"
      >
        <span
          class="label"
        >
          word-language-selector-btn-label-pl
        </span>
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/pl.png"
            srcset="https://flagcdn.com/w40/pl.png 2x"
            width="20"
          />
        </span>
      </button>
      <button
        class="translation-btn"
        title="result-language-selector-btn-title-en"
      >
        <span
          class="label"
        >
          word-language-selector-btn-label-en
        </span>
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/us.png"
            srcset="https://flagcdn.com/w40/us.png 2x"
            width="20"
          />
        </span>
      </button>
    </div>
  </div>
</div>
`;
